### PR TITLE
[Merged by Bors] - feat: norm_num ext for `Int.floor`

### PR DESCRIPTION
**Pull request by eric-wieser**
I tried this a while ago in https://github.com/leanprover-community/mathlib/pull/16502; this is now just one of the handlers from that PR.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

Some questions:
* Should these be in the `NormNum/` directory?
* Are the `isFoo` lemmas supposed to be in the `Mathlib.Meta.NormNum` namespace?
* Is there any point in the `isNat` and `isInt` handlers, since `simp` can already deal with these?
* ~~Do I really need `CharZero`?~~

### CODE REVIEW

File: test/norm_num_ext.lean [hunk 1]
```diff
@@ -389,6 +389,18 @@ example : ∏ i ∈ Finset.range 9, Nat.sqrt (i + 1) = 96 := by norm_num1
 
 end big_operators
 
+section floor
+
+variables (R : Type*) [LinearOrderedRing R] [FloorRing R]
+variables (K : Type*) [LinearOrderedField K] [FloorRing K]
```
Comment (YaelDillies) [inline test/norm_num_ext.lean:395]: Lean 3-ism

File: Mathlib/Data/Rat/Floor.lean [hunk 2]
```diff
@@ -87,6 +87,48 @@ theorem cast_fract (x : ℚ) : (↑(fract x) : α) = fract (x : α) := by
...
+theorem isInt_intFloor_ofIsRat [CharZero α] (r : α) (n : ℤ) (d : ℕ) :
+    IsRat r n d → IsInt ⌊r⌋ (n / d) := by
+  rintro ⟨inv, rfl⟩
+  constructor
+  simp only [invOf_eq_inv, ← div_eq_mul_inv, Int.cast_id]
+  rw [← floor_int_div_nat_eq_div n d, ← floor_cast (α := α), Rat.cast_div,
+    cast_intCast, cast_natCast]
+
+@[norm_num ⌊_⌋]
+def evalIntFloor : NormNumExt where eval {u αZ} e := do
```
Comment (YaelDillies) [inline Mathlib/Data/Rat/Floor.lean:109]: missing docstring

File: Mathlib/Data/Rat/Floor.lean [hunk 3]
```diff
@@ -87,6 +87,48 @@ theorem cast_fract (x : ℚ) : (↑(fract x) : α) = fract (x : α) := by
...
+
+open Mathlib.Meta.NormNum Qq
+
+theorem isNat_intFloor {R} [LinearOrderedRing R] [FloorRing R] (r : R) (m : ℕ) :
+    IsNat r m → IsNat ⌊r⌋ m := by rintro ⟨⟨⟩⟩; exact ⟨by simp⟩
+
+theorem isInt_intFloor {R} [LinearOrderedRing R] [FloorRing R] (r : R) (m : ℤ) :
+    IsInt r m → IsInt ⌊r⌋ m := by rintro ⟨⟨⟩⟩; exact ⟨by simp⟩
+
+theorem isInt_intFloor_ofIsRat [CharZero α] (r : α) (n : ℤ) (d : ℕ) :
```
Comment (YaelDillies) [inline Mathlib/Data/Rat/Floor.lean:100]: As mentioned on Zulip, this `CharZero` requirement comes for free once you import `Algebra.Order.Ring.CharZero`. I really dislike for basic inheritance typeclasses to be hidden in obscure files, so I have opened #13654.

**Comment by fpvandoorn** [comment]
Marking as `awaiting-author`. Please replace that label with `awaiting-review` if appropriate.

File: Mathlib/Data/Rat/Floor.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Data/Rat/Floor.lean]: Can you merge master now that #13654 has landed?

Comment (eric-wieser) [reply Mathlib/Data/Rat/Floor.lean]: Done

**Review by YaelDillies — APPROVED** [review]
Looking good!

maintainer merge

**Comment by fpvandoorn** [comment]
Thanks!

bors merge

The versions for ceiling and nat-floor/nat-ceil would also be great!

**Pull request closed without merging**